include/seqan3/search/bloom_filter/bloom_filter.hpp

```diff
@@ -0,0 +1,386 @@
+            throw std::logic_error{"The number of hash functions must be > 0 and <= 5."};
+        if (size_ == 0)
+            throw std::logic_error{"The size of a bloom filter must be > 0."};
+
+        hash_shift = std::countl_zero(size_);
+        data = sdsl::bit_vector(size_);
+    }
+
+    /*!\brief Construct a compressed Bloom Filter.
+     * \param[in] bf The uncompressed seqan3::bloom_filter.
+     *
+     * \attention This constructor can only be used to construct **compressed** Bloom Filters.
+     *
+     * \details
+     *
+     * ### Example
+     *
+     * \include test/snippet/search/bloom_filter/bloom_filter_constructor_compressed.cpp
+     */
+    bloom_filter(bloom_filter<data_layout::uncompressed> const & bf)
+    //!\cond
+        requires (data_layout_mode == data_layout::compressed)
+    //!\endcond
+    {
+        std::tie(size_, hash_shift, hash_funs) =
+            std::tie(bf.size_, bf.hash_shift, bf.hash_funs);
+
+        data = sdsl::sd_vector<>{bf.data};
+    }
+    //!\}
+
+    /*!\name Modifiers
+     * \{
+     */
+    /*!\brief Inserts a value into the bloom filter.
+     * \param[in] value The raw numeric value to process.
+     *
+     * \attention This function is only available for **uncompressed** Bloom Filters.
+     *
+     * \details
+     *
+     * ### Example
+     *
+     * \include test/snippet/search/bloom_filter/bloom_filter_emplace.cpp
+     */
+    void emplace(size_t const value) noexcept
+    //!\cond
+        requires (data_layout_mode == data_layout::uncompressed)
+    //!\endcond
+    {
+        for (size_t i = 0; i < hash_funs; ++i)
+        {
+            size_t idx = hash_and_fit(value, hash_seeds[i]);
+            assert(idx < data.size());
+            data[idx] = 1;
+        };
+    }
+
+    /*!\brief Clears the complete Bloom Filter data.
+     *
+     * \attention This function is only available for **uncompressed** Bloom Filters.
+     *
+     * \details
+     *
+     * ### Example
+     *
+     * \include test/snippet/search/bloom_filter/bloom_filter_clear.cpp
+     */
+    void clear() noexcept
+    //!\cond
+        requires (data_layout_mode == data_layout::uncompressed)
+    //!\endcond
+    {
+        data = sdsl::bit_vector(size_);
+    }
+    //!\}
+    
+    /*!\name Lookup
+     * \{
+     */
+    /*!\brief Check whether a value is present in the Bloom Filter.
+     * \param[in] value The raw numeric value to process.
+     *
+     * \attention This function is only available for **uncompressed** Bloom Filters.
+     *
+     * \details
+     *
+     * ### Example
+     *
+     * \include test/snippet/search/bloom_filter/bloom_filter_contains.cpp
+     */
+    bool contains(size_t const value) const noexcept
+    {
+        for ( size_t i = 0; i < hash_funs; i++ )
+        {
+            size_t idx = hash_and_fit(value, hash_seeds[i]);
+            assert(idx < data.size());
+            if ( data[idx] == 0 )
+                return false;
+        }
+        return true;
```

Comment:
Not sure if this is what you meant, but I checked the following with a 4GBit BF and, depending on the hit ratio, 250,000 - 1,000,000 entries:

```c++
// In bloom_filter.hpp
    bool contains(size_t const value) const noexcept
    {
        for ( size_t i = 0; i < hash_funs; i++ )
        {
            size_t idx = hash_and_fit(value, hash_seeds[i]);
            assert(idx < data.size());
            if ( data[idx] == 0 )
                return false;
        }
        return true;
    }

    bool contains_ibf_version(size_t const value) const noexcept
    {

        std::array<size_t, 5> bloom_filter_indices;
        std::memcpy(&bloom_filter_indices, &hash_seeds, sizeof(size_t) * hash_funs);

        for (size_t i = 0; i < hash_funs; ++i)
            bloom_filter_indices[i] = hash_and_fit(value, bloom_filter_indices[i]);

        for (size_t i = 0; i < hash_funs; ++i)
        {
            assert(bloom_filter_indices[i] < data.size());
            if ( data[bloom_filter_indices[i]] == 0 )
                return false;
        }

        return true;
    }
```

```c++
// In main executable
    size_t BF_SIZE = 4000000000;
    size_t BF_ENTRIES = 1000000;

    // Check use of 2-5 hash functions
    for ( size_t hash_funs = 2; hash_funs < 6; ++hash_funs)
    {
        std::cout << std::to_string(hash_funs) << " hash functions:\n";
        
        // Check hit "probability" of 0.5, 0.66, 0.75 and 1.0 
        for ( size_t prob_divider = 1; prob_divider <= 4; prob_divider++ )
        {
            std::cout << "   Hit probability of " << std::to_string(1.0f/prob_divider) << "\n";
            seqan3::bloom_filter bf(seqan3::bin_size{BF_SIZE}, seqan3::hash_function_count{hash_funs});
            for ( size_t hash = 0; hash < BF_ENTRIES; hash += 1 )
            {
                // Only emplace every 1st, 2nd, 3rd, or 4th entry, based on the prob_divider
                if ( hash % prob_divider == 0 )
                    bf.emplace(hash);
            }

            // Measure with BF version
            auto start = std::chrono::high_resolution_clock::now();

            size_t hits = 0;
            for ( size_t hash = 0; hash < BF_ENTRIES; hash += 1 )
                hits += bf.contains(hash);

            auto finish = std::chrono::high_resolution_clock::now();
            std::chrono::duration<double> elapsed = finish - start;            
            std::cout << "      BF version: Found " << hits << " hits in " << elapsed.count() << " s\n";

            // Measure with IBF version
            start = std::chrono::high_resolution_clock::now();

            hits = 0;
            for ( size_t hash = 0; hash < BF_ENTRIES; hash += 1 )
                hits += bf.contains_ibf_version(hash);

            finish = std::chrono::high_resolution_clock::now();
            elapsed = finish - start;            
            std::cout << "      IBF version: Found " << hits << " hits in " << elapsed.count() << " s\n";
        }
    }
```

Not sure whether I messed something up with the benchark (please check), but compiled in release mode I obtain the following results:
```
2 hash functions:
   Hit probability of 1.000000
      BF version: Found 1000000 hits in 0.282854 s
      IBF version: Found 1000000 hits in 0.322944 s
   Hit probability of 0.500000
      BF version: Found 500000 hits in 0.19024 s
      IBF version: Found 500000 hits in 0.292299 s
   Hit probability of 0.333333
      BF version: Found 333334 hits in 0.160255 s
      IBF version: Found 333334 hits in 0.306781 s
   Hit probability of 0.250000
      BF version: Found 250000 hits in 0.151015 s
      IBF version: Found 250000 hits in 0.358289 s
3 hash functions:
   Hit probability of 1.000000
      BF version: Found 1000000 hits in 0.351634 s
      IBF version: Found 1000000 hits in 0.568406 s
   Hit probability of 0.500000
      BF version: Found 500000 hits in 0.249056 s
      IBF version: Found 500000 hits in 0.404341 s
   Hit probability of 0.333333
      BF version: Found 333334 hits in 0.197615 s
      IBF version: Found 333334 hits in 0.389003 s
   Hit probability of 0.250000
      BF version: Found 250000 hits in 0.181299 s
      IBF version: Found 250000 hits in 0.411997 s
4 hash functions:
   Hit probability of 1.000000
      BF version: Found 1000000 hits in 0.515441 s
      IBF version: Found 1000000 hits in 0.627472 s
   Hit probability of 0.500000
      BF version: Found 500000 hits in 0.296285 s
      IBF version: Found 500000 hits in 0.509962 s
   Hit probability of 0.333333
      BF version: Found 333334 hits in 0.238565 s
      IBF version: Found 333334 hits in 0.453795 s
   Hit probability of 0.250000
      BF version: Found 250000 hits in 0.214064 s
      IBF version: Found 250000 hits in 0.44655 s
5 hash functions:
   Hit probability of 1.000000
      BF version: Found 1000000 hits in 0.590384 s
      IBF version: Found 1000000 hits in 0.845794 s
   Hit probability of 0.500000
      BF version: Found 500000 hits in 0.356562 s
      IBF version: Found 500000 hits in 0.628335 s
   Hit probability of 0.333333
      BF version: Found 333334 hits in 0.279993 s
      IBF version: Found 333334 hits in 0.533912 s
   Hit probability of 0.250000
      BF version: Found 250000 hits in 0.235909 s
      IBF version: Found 250000 hits in 0.489913 s
```

Thus, at least in this setup, the BF is **significantly** faster in every of the chosen setups. I am not sure why the difference is even that big for 100% hits, but maybe the initialization of the array + memcpy takes actually more time than "switching context"!?

Reply:
Seems like a missed optimization / premature "optimzation" on our part.
Thanks for testing :)

Reply:
I think this is not quite comparable, since the BF version drops out of the function ones it found a 0 bit, doesn't it?
But the IBF needs to compute all hashes and do the binary-and operation. One could check for a 0 binning bit vector and terminate early, as well. But this might only be useful if there are actually more queries that are not contained in any bin, I presume? Would it be possible to also try the following version? Just wondering, how this compares to the other when the hit probabilty is 100% (at least uses only one for loop and does not need to initialise the array before, but otherwise should have the same behavior, that is examining all hash functions)
```cpp
    bool contains_ibf_version_2(size_t const value) const noexcept
    {
        bool result = true;
        for ( size_t i = 0; i < hash_funs; ++i)
        {
            size_t idx = hash_and_fit(value, hash_seeds[i]);
            assert(idx < data.size());
            result &= data[idx];
            // could also be just one line:  result &= data[hash_and_fit(value, hash_seeds[i])];
        }
        return result;
    }
```

Reply:
> I think this is not quite comparable, since the BF version drops out of the function once it found a 0 bit, doesn't it?
> But the IBF needs to compute all hashes and do the binary-and operation. 

Yes, that's why I originally changed the implementation. I thought that @eseiler actually wanted to see whether the influence of memory optimization might be higher than breaking hash value computation. 

I just realized that I actually might have messed something up in the previous benchmark where I had much higher runtimes than now ... I first thought that I maybe mistyped something in the compile parameters making it run in Debug, but even then I get very different numbers. Not sure what was going on there, probably my local setup is simply not ideal for benchmarking (at least, in terms of being comparable between different sessions) ...

Well, however, the key tendencies are still the same as before, though for smaller hit ratios, the difference in the 100% Hits is a bit lower now. Running with 10M hits (10 times more than in the benchmarks showed above, to compensate for the much smaller time stamps), I get the following results. IBF version 3 is the same as IBF version, but adding the hash calculation to the second for loop to also interrupt hash value calculation in that version (though, we are changing context again in this case, so I wouldn't expect it to be really useful in practice):

```
2 hash functions:
   Hit probability of 1.000000
      BF version:    Found 10000000 hits in 0.494258 s
      IBF version:   Found 10000000 hits in 0.564766 s
      IBF version 2: Found 10000000 hits in 0.516285 s
      IBF version 3: Found 10000000 hits in 0.515815 s
   Hit probability of 0.500000
      BF version:    Found 5000036 hits in 0.353638 s
      IBF version:   Found 5000036 hits in 0.434797 s
      IBF version 2: Found 5000036 hits in 0.491515 s
      IBF version 3: Found 5000036 hits in 0.412789 s
   Hit probability of 0.333333
      BF version:    Found 3333349 hits in 0.345686 s
      IBF version:   Found 3333349 hits in 0.395406 s
      IBF version 2: Found 3333349 hits in 0.515044 s
      IBF version 3: Found 3333349 hits in 0.379747 s
   Hit probability of 0.250000
      BF version:    Found 2500013 hits in 0.227909 s
      IBF version:   Found 2500013 hits in 0.426553 s
      IBF version 2: Found 2500013 hits in 0.551206 s
      IBF version 3: Found 2500013 hits in 0.369852 s
3 hash functions:
   Hit probability of 1.000000
      BF version:    Found 10000000 hits in 0.790269 s
      IBF version:   Found 10000000 hits in 0.776725 s
      IBF version 2: Found 10000000 hits in 0.783915 s
      IBF version 3: Found 10000000 hits in 0.831222 s
   Hit probability of 0.500000
      BF version:    Found 5000001 hits in 0.533165 s
      IBF version:   Found 5000001 hits in 0.590132 s
      IBF version 2: Found 5000001 hits in 0.748815 s
      IBF version 3: Found 5000001 hits in 0.510539 s
   Hit probability of 0.333333
      BF version:    Found 3333335 hits in 0.452182 s
      IBF version:   Found 3333335 hits in 0.624311 s
      IBF version 2: Found 3333335 hits in 0.751224 s
      IBF version 3: Found 3333335 hits in 0.484645 s
   Hit probability of 0.250000
      BF version:    Found 2500000 hits in 0.391202 s
      IBF version:   Found 2500000 hits in 0.56284 s
      IBF version 2: Found 2500000 hits in 0.775221 s
      IBF version 3: Found 2500000 hits in 0.408564 s
4 hash functions:
   Hit probability of 1.000000
      BF version:    Found 10000000 hits in 1.02324 s
      IBF version:   Found 10000000 hits in 1.11717 s
      IBF version 2: Found 10000000 hits in 0.993771 s
      IBF version 3: Found 10000000 hits in 1.06993 s
   Hit probability of 0.500000
      BF version:    Found 5000000 hits in 0.62999 s
      IBF version:   Found 5000000 hits in 0.78361 s
      IBF version 2: Found 5000000 hits in 1.00452 s
      IBF version 3: Found 5000000 hits in 0.66989 s
   Hit probability of 0.333333
      BF version:    Found 3333334 hits in 0.501071 s
      IBF version:   Found 3333334 hits in 0.686884 s
      IBF version 2: Found 3333334 hits in 0.964291 s
      IBF version 3: Found 3333334 hits in 0.526065 s
   Hit probability of 0.250000
      BF version:    Found 2500000 hits in 0.443009 s
      IBF version:   Found 2500000 hits in 0.641082 s
      IBF version 2: Found 2500000 hits in 0.891113 s
      IBF version 3: Found 2500000 hits in 0.484015 s
5 hash functions:
   Hit probability of 1.000000
      BF version:    Found 10000000 hits in 1.05258 s
      IBF version:   Found 10000000 hits in 1.27367 s
      IBF version 2: Found 10000000 hits in 1.17618 s
      IBF version 3: Found 10000000 hits in 1.28749 s
   Hit probability of 0.500000
      BF version:    Found 5000000 hits in 0.689612 s
      IBF version:   Found 5000000 hits in 1.03318 s
      IBF version 2: Found 5000000 hits in 1.19449 s
      IBF version 3: Found 5000000 hits in 0.784504 s
   Hit probability of 0.333333
      BF version:    Found 3333334 hits in 0.574388 s
      IBF version:   Found 3333334 hits in 0.845826 s
      IBF version 2: Found 3333334 hits in 1.1951 s
      IBF version 3: Found 3333334 hits in 0.606484 s
   Hit probability of 0.250000
      BF version:    Found 2500000 hits in 0.508206 s
      IBF version:   Found 2500000 hits in 0.890143 s
      IBF version 2: Found 2500000 hits in 1.22659 s
      IBF version 3: Found 2500000 hits in 0.555937 s
```

Thus, the game changer is definitely breaking the loop. I am not sure how expensive a comparison to an empty binning bit vector would be, and if we would expect to actually have this situation very often in the typical applications of the IBF, but without the loop breaking I think the difference between the versions is too small (and probably depends on hardware, system, application and data) to really care about it.

Reply:
What is ibf version 3? 
Just for my understanding, when I have 100 % hit probability, then I never break the loop? So the number of loop iterations should be always the same in all versions of this test case? Would be interesting to look into the generated assembly of the different versions.

Reply:
> What is ibf version 3?

> IBF version 3 is the same as IBF version, but adding the hash calculation to the second for loop to also interrupt hash value calculation in that version

Sorry, was too lazy to put the code, I thought that short sentence would be clear enough.

```c++
    bool contains_ibf_version_3(size_t const value) const noexcept
    {
        std::array<size_t, 5> bloom_filter_indices;
        std::memcpy(&bloom_filter_indices, &hash_seeds, sizeof(size_t) * hash_funs);
        
        for (size_t i = 0; i < hash_funs; ++i)
        {
            bloom_filter_indices[i] = hash_and_fit(value, bloom_filter_indices[i]);
            assert(bloom_filter_indices[i] < data.size());
            if ( data[bloom_filter_indices[i]] == 0 )
                return false;
        }

        return true;
    }
```

> Just for my understanding, when I have 100 % hit probability, then I never break the loop? So the number of loop iterations should be always the same in all versions of this test case?

Yes. That's why there are only minor differences for 100% hit prob.

> Would be interesting to look into the generated assembly of the different versions.

I am really not familiar with this, I am not sure how to do this properly and whether I would actually be able to understand it.

Reply:
Ah sorry. I haven't seen that you already explained it. My bad. Ok  so basically IBF3 is like BF but with the overhead of doing unnecessary memcpy before. But ok, then this is fine enough. With the assembly I was just curious, what happens under the hood. Nevermind. But then the BF version it is :) Thanks for exploring all this.